File: clients/php/Model/Vec2.php
<?php

namespace Model {
    require_once 'Stream.php';

    /**
     * 2 dimensional vector.
     */
    class Vec2
    {
        /**
         * `x` coordinate of the vector
         */
        public float $x;
        /**
         * `y` coordinate of the vector
         */
        public float $y;
    
        function __construct(float $x, float $y)
        {
            $this->x = $x;
            $this->y = $y;
        }
    
        /**
         * Read Vec2 from input stream
         */
        public static function readFrom(\InputStream $stream): Vec2
        {
            $x = $stream->readDouble();
            $y = $stream->readDouble();
            return new Vec2($x, $y);
        }
        
        /**
         * Write Vec2 to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeDouble($this->x);
            $stream->writeDouble($this->y);
        }
    }
}

File: clients/cpp/model/Vec2.cpp
#include "Vec2.hpp"

namespace model {

Vec2::Vec2(double x, double y) : x(x), y(y) { }

// Read Vec2 from input stream
Vec2 Vec2::readFrom(InputStream& stream) {
    double x = stream.readDouble();
    double y = stream.readDouble();
    return Vec2(x, y);
}

// Write Vec2 to output stream
void Vec2::writeTo(OutputStream& stream) const {
    stream.write(x);
    stream.write(y);
}

// Get string representation of Vec2
std::string Vec2::toString() const {
    std::stringstream ss;
    ss << "Vec2 { ";
    ss << "x: ";
    ss << x;
    ss << ", ";
    ss << "y: ";
    ss << y;
    ss << " }";
    return ss.str();
}

}